Convert `mem::zeroed()` / `0` to `MaybeUninit::uninit()`

MarijnS95 opened the pull request:
As noted in #792 changes like this might help us more strictly identify and validate that the argument in question is only relevant as an output argument (i.e. structs with `sType` are read by the function call, even if the caller strictly expects return values there, so that it can fill in multiple structures in a `pNext` chain, and must hence be `Default`- initialized).

It does embed rather nicely with the `assume_init_on_success()` API, though.

---

Creating this as draft just to feel out where we stand with this; it might not net us too much, and doesn't seem like it would help us audit for accidental `sType` structs that are zero/uninit.  There was also the idea to initialize handles with `::null()` explicitly (instead of `mem::zeroed()` or the `MaybeUninit::uninit()` added here), but that also doesn't sound like much of an improvement.

Ralith commented on ash/src/device.rs:
```diff
@@ -828,14 +832,15 @@ impl Device {
     ) -> usize {
-        let mut count = 0;
+        // TODO: when pSparseMemoryRequirements is NULL, is it okay/safe for count to be uninit?
+        let mut count = mem::MaybeUninit::uninit();
```
I think this is sound as rwitten. The spec seems to be missing the relevant text, but consider this analogous section:

> **If pPhysicalDevices is NULL**, then the number of physical devices available is returned in pPhysicalDeviceCount. Otherwise, pPhysicalDeviceCount must point to a variable set by the user to the number of elements in the pPhysicalDevices array [...]

MarijnS95 replied:
Oh, I thought I had deleted all TODOs from this PR again :sweat_smile:

MarijnS95 replied:
@Ralith I'm marking this as unresolved because, at least for `vkGetLatencyTimingsNV` (https://github.com/ash-rs/ash/issues/814, https://github.com/KhronosGroup/Vulkan-Docs/issues/2269#issuecomment-1813339455), `pTimingCount` will be set to the available number of timings _if it was `0`_, even if `pTimings` **was non-`NULL`**. Meaning that there's probably an `if (!pTimings || !*pTimingCount)` in the driver which could read uninitialized values.

@chansen-nv is adding that mention to the spec, but I don't think this behaviour is consistent/expected wrt other Vulkan functions?

---

However, https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceImageSparseMemoryRequirements.html only says:

> `pSparseMemoryRequirementCount` is a pointer to an integer related to the number of sparse memory requirements available or queried, as described below.

(There is no description below)

And the VUID doesn't rule out a similar implementation to `vkGetLatencyTimingsNV` either:

> If the value referenced by `pSparseMemoryRequirementCount` is not `0`, and `pSparseMemoryRequirements` is not `NULL`, `pSparseMemoryRequirements` must be a valid pointer to an array of `pSparseMemoryRequirementCount` [VkSparseImageMemoryRequirements2](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSparseImageMemoryRequirements2.html) structures

Perhaps @oddhack can help us clarify whether or not the count pointer/value should be written when it's zero even if the array-pointer is `NULL` (in the generic case). If we cannot get a guarantee on this for some/most functions, we should back this change out and keep a `0` init (`let mut count = 0;`).

Ralith reviewed (approved):
I agree that this helps clarify whether an intention to pass in a specific value exists or not. Perhaps a marginal change, but certainly doesn't hurt.

MarijnS95 commented:
Sure thing! I'll get going on finding the remaining assignments for `0` / `mem::zeroed()`...

MarijnS95 commented:
@Ralith would you mind taking another look? Besides tackling the remaining obvious `= 0;` and `mem::zeroed()`, I've also converted the `let fd = -1;` initializations to this (assuming Vulkan always initializes them when it retrurns `OK`), and replaced a few remaining zero-initialized `Vec`s with `Vec::with_capacity()` + `set_len()`.

Ralith commented on ash/src/extensions/amdx/shader_enqueue.rs:
```diff
@@ -38,7 +38,9 @@ impl ShaderEnqueue {
         )
-        .result_with_success(pipelines)
+        .result()?;
+        pipelines.set_len(create_infos.len());
```
Should we have a `set_len_on_success<T>(self, v: Vec<T>, len: usize) -> Result<Vec<T>>` helper?

MarijnS95 replied:
Sounds like a good idea, maybe we can share some code with other vector initializers, I'll have a look at that. Even cooler would be something utilizing `Vec<MaybeUninit<T>>` but `with_capacity()` serves the same purpose for now.

Ralith replied:
> Even cooler would be something utilizing Vec<MaybeUninit<T>>

I think that's specifically redundant to `Vec`'s notion of capacity, yeah. We could do `Box<[MaybeUninit<T>]>` and convert that to `Box<[T]>` but that seems needlessly esoteric.

MarijnS95 replied:
Fair point, there are even helper functions to get the uninitialized part of a `Vec` as `&mut [MaybeUninit<T>]` but that's not really of use to us.

Note that a similar call for pipelines in `device.rs` was returning this `set_len`'d `Vec` even when `create_graphics_pipelines()`/`create_compute_pipelines()` returns `Err`. This is because some of the pipelines could have been created, but I don't remember if the driver sets the rest to `VK_NULL_HANDLE` or if we should initialize it to that.

Ralith reviewed (approved):
LGTM. Didn't double-check the spec for every single command, but each case got at least a quick look.
